syntax = "proto3";

package anolis.deviceprovider.v1;

import "google/protobuf/timestamp.proto";

// -----------------------------
// Envelope messages
// -----------------------------

message Request {
  uint64 request_id = 1;

  oneof payload {
    HelloRequest hello = 10;
    ListDevicesRequest list_devices = 11;
    DescribeDeviceRequest describe_device = 12;
    ReadSignalsRequest read_signals = 13;
    CallRequest call = 14;
    GetHealthRequest get_health = 15;
    WaitReadyRequest wait_ready = 16;
  }
}

message Response {
  uint64 request_id = 1;
  Status status = 2;

  oneof payload {
    HelloResponse hello = 10;
    ListDevicesResponse list_devices = 11;
    DescribeDeviceResponse describe_device = 12;
    ReadSignalsResponse read_signals = 13;
    CallResponse call = 14;
    GetHealthResponse get_health = 15;
    WaitReadyResponse wait_ready = 16;
  }
}

// -----------------------------
// Status
// -----------------------------

message Status {
  enum Code {
    CODE_UNSPECIFIED = 0;
    CODE_OK = 1;

    CODE_INVALID_ARGUMENT = 10;
    CODE_NOT_FOUND = 11;
    CODE_FAILED_PRECONDITION = 12;
    CODE_OUT_OF_RANGE = 13;
    CODE_UNIMPLEMENTED = 14;

    CODE_DEADLINE_EXCEEDED = 20;
    CODE_UNAVAILABLE = 21;
    CODE_RESOURCE_EXHAUSTED = 22;

    CODE_INTERNAL = 30;
    CODE_DATA_LOSS = 31;
  }

  Code code = 1;
  string message = 2;
  map<string, string> details = 3;
}

// -----------------------------
// Handshake
// -----------------------------

message HelloRequest {
  string protocol_version = 1;
  string client_name = 2;
  string client_version = 3;
}

message HelloResponse {
  string protocol_version = 1;
  string provider_name = 2;
  string provider_version = 3;
  map<string, string> metadata = 4;
}

// -----------------------------
// Inventory
// -----------------------------

message ListDevicesRequest {
  bool include_health = 1;
}

message ListDevicesResponse {
  repeated Device devices = 1;
  repeated DeviceHealth device_health = 2;
}

message DescribeDeviceRequest {
  string device_id = 1;
}

message DescribeDeviceResponse {
  Device device = 1;
  CapabilitySet capabilities = 2;
}

// -----------------------------
// Telemetry
// -----------------------------

message ReadSignalsRequest {
  string device_id = 1;
  repeated string signal_ids = 2;
  google.protobuf.Timestamp min_timestamp = 3;
}

message ReadSignalsResponse {
  string device_id = 1;
  repeated SignalValue values = 2;
}

// -----------------------------
// Calls
// -----------------------------

message CallRequest {
  string device_id = 1;
  uint32 function_id = 2;
  string function_name = 3;
  map<string, Value> args = 4;
  string idempotency_key = 5;
  google.protobuf.Timestamp deadline = 6;
}

message CallResponse {
  string device_id = 1;
  map<string, Value> results = 2;
  string operation_id = 3;
}

// -----------------------------
// Health
// -----------------------------

message GetHealthRequest {}

message GetHealthResponse {
  ProviderHealth provider = 1;
  repeated DeviceHealth devices = 2;
}

// -----------------------------
// Provider Ready State
// -----------------------------

message WaitReadyRequest {
  uint32 max_wait_ms_hint = 1;  // Optional: client timeout hint
}

message WaitReadyResponse {
  map<string, string> diagnostics = 1;  // init_time_ms, device_count, firmware, etc.
}

// -----------------------------
// Shared types
// -----------------------------

message Device {
  string device_id = 1;
  string provider_name = 2;
  string type_id = 3;
  string type_version = 4;
  string label = 5;
  string address = 6;
  map<string, string> tags = 7;
}

message CapabilitySet {
  repeated FunctionSpec functions = 1;
  repeated SignalSpec signals = 2;
}

message FunctionSpec {
  uint32 function_id = 1;
  string name = 2;
  string description = 3;
  FunctionPolicy policy = 4;
  repeated ArgSpec args = 5;
  repeated ArgSpec results = 6;
}

message FunctionPolicy {
  enum Category {
    CATEGORY_UNSPECIFIED = 0;
    CATEGORY_READ = 1;
    CATEGORY_CONFIG = 2;
    CATEGORY_ACTUATE = 3;
  }

  Category category = 1;
  repeated string allowed_modes = 2;
  bool requires_lease = 3;
  string safety_profile = 4;
  uint32 min_interval_ms = 5;
  bool is_idempotent = 6;
}

message SignalSpec {
  string signal_id = 1;
  string name = 2;
  string description = 3;
  ValueType value_type = 4;
  string unit = 5;
  double poll_hint_hz = 6;
  uint32 stale_after_ms = 7;
}

message SignalValue {
  string signal_id = 1;
  Value value = 2;
  google.protobuf.Timestamp timestamp = 3;
  Quality quality = 4;
  map<string, string> metadata = 5;

  enum Quality {
    QUALITY_UNSPECIFIED = 0;
    QUALITY_OK = 1;
    QUALITY_STALE = 2;
    QUALITY_FAULT = 3;
    QUALITY_UNKNOWN = 4;
  }
}

message ProviderHealth {
  enum State {
    STATE_UNSPECIFIED = 0;
    STATE_OK = 1;
    STATE_DEGRADED = 2;
    STATE_FAULT = 3;
  }

  State state = 1;
  string message = 2;
  map<string, string> metrics = 3;
}

message DeviceHealth {
  string device_id = 1;

  enum State {
    STATE_UNSPECIFIED = 0;
    STATE_OK = 1;
    STATE_STALE = 2;
    STATE_UNREACHABLE = 3;
    STATE_FAULT = 4;
  }

  State state = 2;
  string message = 3;
  google.protobuf.Timestamp last_seen = 4;
  map<string, string> metrics = 5;
}

message ArgSpec {
  string name = 1;
  ValueType type = 2;
  string unit = 3;
  string description = 4; 
  bool required = 5;
  double min_double = 6;
  double max_double = 7;
  int64 min_int64 = 8;
  int64 max_int64 = 9;
  uint64 min_uint64 = 10;
  uint64 max_uint64 = 11;
  repeated string allowed_values = 12;  // Enum constraints for STRING type only
}

enum ValueType {
  VALUE_TYPE_UNSPECIFIED = 0;
  VALUE_TYPE_BOOL = 1;
  VALUE_TYPE_INT64 = 2;
  VALUE_TYPE_UINT64 = 3;
  VALUE_TYPE_DOUBLE = 4;
  VALUE_TYPE_STRING = 5;
  VALUE_TYPE_BYTES = 6;
}

message Value {
  ValueType type = 1;

  oneof kind {
    bool bool_value = 2;
    int64 int64_value = 3;
    uint64 uint64_value = 4;
    double double_value = 5;
    string string_value = 6;
    bytes bytes_value = 7;
  }
}
